packages/ifpack2/src/Ifpack2_DatabaseSchwarz_decl.hpp

```diff
@@ -0,0 +1,402 @@
+// getCrsMatrix.  I would very much prefer that this method not exist.
+// Furthermore, it is both unsafe (MatrixType need not be CrsMatrix)
+// and completely redundant (just call getMatrix() and do the
+// dynamic_cast yourself).
+#include "Teuchos_SerialDenseMatrix.hpp"
+#include "Tpetra_CrsMatrix_decl.hpp"
+
+#include <type_traits>
+
+namespace Ifpack2 {
+
+/// \class DatabaseSchwarz
+/// \brief Overlapping Schwarz where redundant patches are not stored explicitly.
+/// \tparam MatrixType The type of matrix to use.
+///
+/// This class implements an overlapping Schwarz assuming it has already
+/// been provided overlapping data containers. Additionally, it is assumed
+/// there is at least one row with PatchSize_ nonzeros for each patch.
+/// This means that a matrix obtained from piecewise linear FEMs in 2d,
+/// for example, will not have rows with 4 nonzeros for each patch.
+/// However, piecewise quadratic FEMs in 2d will have one row with 9 nonzeros
+/// for each cell in the mesh, meaning all the patches will be detected.
+/// For a matrix corresponding to a Taylor-Hood discretization in 2d
+/// (quadratic velocities and linear pressures), the patch size would be 22.
+///
+/// The general algorithm proceeds as follows:
+/// <ol>
+/// <li> The rows of A are analyzed sequentially for any row with num_entries == PatchSize_ </li>
+/// <li> If the current row corresponds to a DOF that has already been "visited", it is skipped </li>
+/// <li> Then, all nonzero indices belonging to the row are marked as "visited" </li>
+/// <li> The local patch matrix is formed and compared to a database of previous patch matrices.
+///      If any patch matrix in the database has an l1 distance to the current patch matrix less than tol,
+///      the current patch is not matrix is not stored, but an index pointing it to a replacement is stored instead </li>
+/// <li> Finally, if this patch matrix is not sufficiently close to one that has already been seen, it is added to the database </li>
+/// <li> The compute phase then inverts only the patch matcies in the database,
+///      and the apply phase loops over all patches and applies the inverse of each appropriate patch matrix </li>
+/// </ol>
+///
+/// In general, there is a noticeable speedup when using this method
+/// compared to a typical method. This speedup may be further improved
+/// by using more advanced linear algebra interfaces such as batched
+/// Kokkos solves instead of the current LAPACK approach.
+template<class MatrixType>
+class DatabaseSchwarz :
+    virtual public Ifpack2::Preconditioner<typename MatrixType::scalar_type,
+                                           typename MatrixType::local_ordinal_type,
+                                           typename MatrixType::global_ordinal_type,
+                                           typename MatrixType::node_type>,
+    virtual public Ifpack2::Details::CanChangeMatrix<Tpetra::RowMatrix<typename MatrixType::scalar_type,
```

Comment:
Would it make sense to derive from Ifpack2::AdditiveSchwarz (or whatever this is based on) instead of duplicating?

Reply:
The options I looked at were 
 1. Add this as another option in `AdditiveSchwarz`.
 2. Derive from `AdditiveSchwarz`.
 3. Derive from `Preconditioner`, and use it as an inner solve with `AdditiveSchwarz`.

For 1., it seemed like this main idea of constructing a database was large enough and required enough code and data structures that this wouldn't make sense. For 2., I could do something like that and then force this to be the inner solve every time, but there would be weird conflicts like having to rewrite the parameterlist settings with the new name and then trying to explicitly call the parent class methods first anyway. I felt like 3. was the easiest route since we already had options to handle inner preconditioners with `AdditiveSchwarz`, so I wouldn't have to manually juggle any parent-child interactions/reimplementations/overrides/etc.